cmd/juju/application/upgradecharm.go

```diff
@@ -220,9 +218,6 @@ func (c *upgradeCharmCommand) SetFlags(f *gnuflag.FlagSet) {
 func (c *upgradeCharmCommand) Init(args []string) error {
 	switch len(args) {
 	case 1:
-		if !names.IsValidApplication(args[0]) {
```

Comment:
Sorry wrote this before coffee. I grok your explanation better, but I don't think my stance has changed. Even if we communicate using what's passed in, I think the controller needs to have the ultimate say.

E.g. (w/ made up versions):

- v1 client, v2 controller, client-side validation: client cannot deploy charms with v2 controller tags.
- v1 client, v2 controller, no client-side validation: client can deploy v2 controller tags and v1 tags if the controller has backwards compatibility.
- v2 client, v1 controller, client-side validation: client cannot deploy charms with v1 controller tags.
- v2 client, v1 controller, no client-side validation: client can deploy v1 controller tags.